fdbclient/FileBackupAgent.actor.cpp

```diff
@@ -458,8 +475,417 @@ Value makePadding(int size) {
+		if (!(header.cipherTextDetails.baseCipherId == textCipherKey->getBaseCipherId() &&
+		      header.cipherTextDetails.encryptDomainId == textCipherKey->getDomainId() &&
+		      header.cipherTextDetails.salt == textCipherKey->getSalt())) {
+			TraceEvent(SevWarn, "EncryptionHeader_CipherTextMismatch")
+			    .detail("TextDomainId", textCipherKey->getDomainId())
+			    .detail("ExpectedTextDomainId", header.cipherTextDetails.encryptDomainId)
+			    .detail("TextBaseCipherId", textCipherKey->getBaseCipherId())
+			    .detail("ExpectedTextBaseCipherId", header.cipherTextDetails.baseCipherId)
+			    .detail("TextSalt", textCipherKey->getSalt())
+			    .detail("ExpectedTextSalt", header.cipherTextDetails.salt);
+			throw encrypt_header_metadata_mismatch();
+		}
+	}
+
+	ACTOR static Future<StringRef> decryptImpl(Database cx,
+	                                           StringRef headerS,
+	                                           const uint8_t* dataP,
+	                                           int64_t dataLen,
+	                                           Arena* arena) {
+		Reference<AsyncVar<ClientDBInfo> const> dbInfo = cx->clientInfo;
+		state BlobCipherEncryptHeader header = BlobCipherEncryptHeader::fromStringRef(headerS);
+		TextAndHeaderCipherKeys cipherKeys = wait(getEncryptCipherKeys(dbInfo, header, BlobCipherMetrics::BACKUP));
+		ASSERT(cipherKeys.cipherHeaderKey.isValid() && cipherKeys.cipherTextKey.isValid());
+		validateEncryptionHeader(cipherKeys.cipherHeaderKey, cipherKeys.cipherTextKey, header);
+		DecryptBlobCipherAes256Ctr decryptor(
+		    cipherKeys.cipherTextKey, cipherKeys.cipherHeaderKey, header.iv, BlobCipherMetrics::BACKUP);
+		return decryptor.decrypt(dataP, dataLen, header, *arena)->toStringRef();
+	}
+
+	static Future<StringRef> decrypt(Database cx,
+	                                 StringRef headerS,
+	                                 const uint8_t* dataP,
+	                                 int64_t dataLen,
+	                                 Arena* arena) {
+		return decryptImpl(cx, headerS, dataP, dataLen, arena);
+	}
+
+	void encrypt() {
+		ASSERT(cipherKeys.headerCipherKey.isValid() && cipherKeys.textCipherKey.isValid());
+		EncryptBlobCipherAes265Ctr encryptor(cipherKeys.textCipherKey,
+		                                     cipherKeys.headerCipherKey,
+		                                     cipherKeys.ivRef.begin(),
+		                                     AES_256_IV_LENGTH,
+		                                     ENCRYPT_HEADER_AUTH_TOKEN_MODE_SINGLE,
+		                                     BlobCipherMetrics::BACKUP);
+		Arena arena;
+		int64_t payloadSize = wPtr - dataPayloadStart;
+		auto encryptedData = encryptor.encrypt(dataPayloadStart, payloadSize, encryptHeader, arena);
+
+		// re-write encrypted data to buffer
+		std::memcpy(dataPayloadStart, encryptedData->begin(), payloadSize);
+	}
+
+	ACTOR static Future<Void> updateEncryptionKeys(EncryptedRangeFileWriter* self,
```

Comment:
The code support two types of encyrption keys:
1. Revocable - cipherKey with `limited` validity
2. Non-revocable - cipherKey with no-expiry, however, may be `refresh`.

Refer to `BlobCipher.h` below routines:

```
	inline bool needsRefresh() {
		if (refreshAtTS == std::numeric_limits<int64_t>::max()) {
			return false;
		}
		return now() >= refreshAtTS ? true : false;
	}

	inline bool isExpired() {
		if (expireAtTS == std::numeric_limits<int64_t>::max()) {
			return false;
		}
		return now() >= expireAtTS ? true : false;
	}
```

It is possible that a given encryptionKey validity (refreshAt or expireAt) goes stale while preparing a `block` to flush; imo, it is OK not to handle that scenario. However, it is highly desirable to check encryption keys `validity` across `blocks` and refresh the encryption key(s) as needed.